src/hotspot/share/gc/shared/barrierSet.cpp

```diff
@@ -57,8 +57,11 @@ static BarrierSetNMethod* select_barrier_set_nmethod(BarrierSetNMethod* barrier_
     // The GC needs nmethod entry barriers to do concurrent GC
     return barrier_set_nmethod;
   } else {
-    // The GC needs nmethod entry barriers to deal with continuations
-    // and code cache unloading
+    // The GC needs nmethod entry barriers for code cache unloading.
+    // Concurrent GC needs them also for preserving the weakly referenced objects in
+    // the constant pool of nmethods as part of the SATB snapshot, and to deal with
+    // compiled frames contained in continuation stack chunks allocated after
+    // concurent mark start.
```

Comment:
I've adapted the comment after https://mail.openjdk.org/pipermail/hotspot-gc-dev/2023-January/041194.html

Reply:
I feel the comment is at the wrong level of abstraction, compared with the surrounding code.

Maybe just `Provide nmethod entry barrier to support concurrent marking and code cache unloading.` at the top of the whole method.

The more elaborate "inside" about nmethod-entry-barrier probably makes more sense at `BarrierSetNMethod::nmethod_entry_barrier`.

Reply:
You are right. I've reverted that change and added a line of comment to `BarrierSetNMethod::nmethod_entry_barrier`